### PR TITLE
fix: `Parse.Query.containedIn` and `matchesQuery` do not work with nested objects

### LINKED ISSUE #7414: Support deeply nested pointers in 'containedIn' queries
### New Issue Checklist

- [x ] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x ] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x ] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x ] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description

Querying using "containedIn" in deep nested objects using pointers doesn't seems to work.

### Steps to reproduce

See example below.

### Actual Outcome

The queries return empty arrays.

### Expected Outcome

We expected an array of Parse.Object .

### Failing Test Case / Pull Request

- [ ] 🤩 I submitted a PR with a fix and a test case.
- [ ] 🧐 I submitted a PR with a failing test case.

To reproduce the issue, I added this at the end of ParseQuery.spec.js

```
  it('deeply nested Pointers (issue #7413)', async function (done) {
    const parent = new Parse.Object('Parent');
    const child1 = new Parse.Object('Child');
    const child2 = new Parse.Object('Child');

    parent.set('children', [
      { child: child1, count: 2 },
      { child: child2, count: 3 },
    ]);
    await parent.save();

    const results = await new Parse.Query('Parent')
      .containedIn('children.child', [child1])
      .find();
    expect(results.length).toBe(1);
    done();
  });
```

This test passes if I replace the line :
`
      .containedIn('children.child', [child1])
`
by
`
      .containedIn('children.child.objectId', [child1.id])
`

###  Environment

parse-server 4.5.0

Server
- Parse Server version: 4.5.0
- Operating system: Windows and debian
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): local Windows and Google App Engine

Database
- System (MongoDB or Postgres): MongoDB
- Database version: 4.4.6
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): local and MongoDB Atlas

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): Javascript
- SDK version: 3.2

### PR DESCRIPTION
**Pull request by RahulLanjewar93**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

## Issue
https://github.com/parse-community/parse-server/issues/7414

Closes: https://github.com/parse-community/parse-server/issues/7414

## Approach
Previously nested objects were converted to `Child$ObjectID` format before making a query to mongodb. Which was incorrect since nested objects are stored in their object format. Current change handles this edge case. Originally even nested objects should have been stored in `Child$ObjectID` format. But this is a fix we can provide for now.

## Tasks

- [x] Add tests

## Summary by CodeRabbit

- **New Features**
  - Improved support for querying nested pointer fields using equalTo, containedIn, and matchesQuery constraints in MongoDB.
- **Tests**
  - Added tests to verify correct behavior of queries involving nested pointer fields with equalTo, containedIn, and matchesQuery constraints.

### CODE REVIEW

**Comment by parseplatformorg** [comment]
### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/acinader/pr-checks/7bea8897-9626-46be-b806-56fb26e61e3b?refs=20f4b73ab96a0fd1668b9261d465fdd4c3a50793&source=prChecksComment)

**Comment by RahulLanjewar93** [comment]
It cancelled the failing test case action -_-
Can someone re run it so that they can see the actual issue

**Comment by mtrezza** [comment]
Tests should be running now

**Comment by RahulLanjewar93** [comment]
> Tests should be running now

i meant the failing test cases were skipped since I pushed a commit before the actions were complete. We need that to see the current behaviour right? Is there a way to retry actions at that specific commit so we can see the failing test cases

**Comment by mtrezza** [comment]
If you say they failed for you when you ran them locally, then that's fine. I see that all tests are passing now as well, nice!

File: spec/MongoStorageAdapter.spec.js [hunk 1]
```diff
@@ -390,6 +390,77 @@ describe_only_db('mongo')('MongoStorageAdapter', () => {
     await expectAsync(adapter.getClass('UnknownClass')).toBeRejectedWith(undefined);
   });
 
+
+  /**
+   * If we use equalTo to comparse the nested pointer it works
+   * But it does not work with contained in or matchesQuery
+   */
+  it('Parse query works with nested objects if equal to is used', async () => {
```
Comment (mtrezza) [inline spec/MongoStorageAdapter.spec.js]:
```suggestion
  it('queries nested objects using equalTo', async () => {
```

File: spec/MongoStorageAdapter.spec.js [hunk 2]
```diff
@@ -390,6 +390,77 @@ describe_only_db('mongo')('MongoStorageAdapter', () => {
...
+    await parent.save();
+
+    const query1 = await new Parse.Query('Parent')
+      .containedIn('some.nested.key.child', [child])
+      .find();
+
+    expect(query1.length).toEqual(1);
+  })
+
+  it('Parse query works when matchesQuery is used which in turn uses contained in', async () => {
```
Comment (mtrezza) [inline spec/MongoStorageAdapter.spec.js]:
```suggestion
  it('queries nested objects using matchesQuery', async () => {
```

File: spec/MongoStorageAdapter.spec.js [hunk 3]
```diff
@@ -390,6 +390,77 @@ describe_only_db('mongo')('MongoStorageAdapter', () => {
...
+    await parent.save();
+
+    const query1 = await new Parse.Query('Parent')
+      .equalTo('some.nested.key.child', child)
+      .find();
+
+    expect(query1.length).toEqual(1);
+  })
+
+  it('Parse query works when containedIn is used', async () => {
```
Comment (mtrezza) [inline spec/MongoStorageAdapter.spec.js]:
```suggestion
  it('queries nested objects using containedIn', async () => {
```

File: src/Adapters/Storage/Mongo/MongoTransform.js [hunk 4]
```diff
@@ -668,18 +671,18 @@ function transformConstraint(constraint, field, count = false) {
   // This is a hack so that:
   //   $regex is handled before $options
   //   $nearSphere is handled before $maxDistance
-  var keys = Object.keys(constraint).sort().reverse();
+  var constraintKeys = Object.keys(constraint).sort().reverse();
```
Comment (mtrezza) [inline src/Adapters/Storage/Mongo/MongoTransform.js]: It seems that most changes are just a var renaming form `keys` to `constraintKeys`. We aim at minimal code changes unless there is a justified benefit (beyond semantics) to rename the var. If you feel the var needs an explanation, then please add a code comment directly to the var which will also be picked up by the IDE:

```js
/** Explain var. */
var keys = ...
```

File: spec/MongoStorageAdapter.spec.js [hunk 5]
```diff
@@ -390,6 +390,77 @@ describe_only_db('mongo')('MongoStorageAdapter', () => {
     await expectAsync(adapter.getClass('UnknownClass')).toBeRejectedWith(undefined);
   });
 
+
+  /**
+   * If we use equalTo to comparse the nested pointer it works
+   * But it does not work with contained in or matchesQuery
+   */
+  it('queries nested objects using equalTo', async () => {
```
Comment (mtrezza) [inline spec/MongoStorageAdapter.spec.js]: I doesn't seem like these tests belong in the `MongoStorageAdapter.spec.js` file, which is testing the adapter itself. These are just Parse.Query tests and there should be a dedicated file for that.

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by mtrezza** [comment]
Note: it seems that your PR is based on the `release` branch of your fork, instead of the `alpha` branch. That doesn't seem to cause any conflicts - surprisingly - but for future PRs, a PR should be based on the same branch as the one it will be merged into.

File: spec/MongoStorageAdapter.spec.js (hunk 5 above)
Comment (RahulLanjewar93) [reply spec/MongoStorageAdapter.spec.js]: I did place it there initially but those tests are also run for postgres I believe? And this changes are mongodb only. 
I will check if there is any place in that file which is dedicated to mongodb

File: src/Adapters/Storage/Mongo/MongoTransform.js (hunk 4 above)
Comment (RahulLanjewar93) [reply src/Adapters/Storage/Mongo/MongoTransform.js]: Ok will revert the changes, i just thought maybe the naming should have been consistent to what we are doing, since the keys are derived from constraint, it should have been constraintKeys. 

Anyways will revert the change

**Comment by RahulLanjewar93** [comment]
> Note: it seems that your PR is based on the `release` branch of your fork, instead of the `alpha` branch. That doesn't seem to cause any conflicts - surprisingly - but for future PRs, a PR should be based on the same branch as the one it will be merged into.

Have made the required changes

File: spec/ParseQuery.spec.js [hunk 6]
```diff
@@ -5306,4 +5306,76 @@ describe('Parse.Query testing', () => {
         expect(score).toEqual([1]);
       }, { useMasterKey: true });
   });
+
+  describe_only_db('mongo')('check if containedIn and matchesQuery works with nested keys', () => {
```
Comment (mtrezza) [inline spec/ParseQuery.spec.js]:
```suggestion
  describe_only_db('mongo')('query nested keys', () => {
```

File: spec/ParseQuery.spec.js [hunk 7]
```diff
@@ -5306,4 +5306,76 @@ describe('Parse.Query testing', () => {
...
+    /**
+     * If we use equalTo to compare the nested pointer it works
+     * But it does not work with contained in or matchesQuery
+     */
+    it('Parse query works with nested objects if equal to is used', async () => {
```
Comment (mtrezza) [inline spec/ParseQuery.spec.js]:
```suggestion
    it('queries nested key using equalTo', async () => {
```

File: spec/ParseQuery.spec.js [hunk 8]
```diff
@@ -5306,4 +5306,76 @@ describe('Parse.Query testing', () => {
...
+      await parent.save();
+  
+      const query1 = await new Parse.Query('Parent')
+        .equalTo('some.nested.key.child', child)
+        .find();
+  
+      expect(query1.length).toEqual(1);
+    });
+  
+    it('Parse query works when containedIn is used', async () => {
```
Comment (mtrezza) [inline spec/ParseQuery.spec.js]:
```suggestion
    it('queries nested key using containedIn', async () => {
```

File: spec/ParseQuery.spec.js [hunk 9]
```diff
@@ -5306,4 +5306,76 @@ describe('Parse.Query testing', () => {
...
+      await parent.save();
+  
+      const query1 = await new Parse.Query('Parent')
+        .containedIn('some.nested.key.child', [child])
+        .find();
+  
+      expect(query1.length).toEqual(1);
+    });
+  
+    it('Parse query works when matchesQuery is used which in turn uses contained in', async () => {
```
Comment (mtrezza) [inline spec/ParseQuery.spec.js]:
```suggestion
    it('queries nested key using matchesQuery', async () => {
```

File: spec/ParseQuery.spec.js [hunk 10]
```diff
@@ -5306,4 +5306,76 @@ describe('Parse.Query testing', () => {
         expect(score).toEqual([1]);
       }, { useMasterKey: true });
   });
+
+  describe_only_db('mongo')('query nested keys', () => {
+    /**
+     * If we use equalTo to compare the nested pointer it works
+     * But it does not work with contained in or matchesQuery
+     */
```
Comment (mtrezza) [inline spec/ParseQuery.spec.js]: I think this comment can removed, it seems to just explain the bug?

Comment (RahulLanjewar93) [reply spec/ParseQuery.spec.js]: Yeah can remove

Comment (RahulLanjewar93) [inline spec/ParseQuery.spec.js]:
```suggestion
```

**Comment by mtrezza** [comment]
This looks good, but the fix for Postgres is still missing, correct? Is there a similar code in the Postgres adapter where the fix can be made, or is that code very different? If it's different and you don't have the knowledge to fix in Postgres, we can merge the fix only for MongoDB for now and create an open issue to fix this also for Postgres at a later time.

**Comment by RahulLanjewar93** [comment]
> This looks good, but the fix for Postgres is still missing, correct? Is there a similar code in the Postgres adapter where the fix can be made, or is that code very different? If it's different and you don't have the knowledge to fix in Postgres, we can merge the fix only for MongoDB for now and create an open issue to fix this also for Postgres at a later time.

I haven't checked it, and I don't have enough knowledge to contribute to postgres code, also stuck with some other priorities currently. I can have a look later when I have some bandwidth or if someone else can pick it up that would be helpful. 

I think lets merge this one for now?

**Comment by mtrezza** [comment]
I think that's fine, also given @dplewis's [comment](https://github.com/parse-community/parse-server/pull/9720#issuecomment-2810398651).

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [8.2.1-alpha.1](https://github.com/parse-community/parse-server/releases/tag/8.2.1-alpha.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [8.2.1](https://github.com/parse-community/parse-server/releases/tag/8.2.1)
